build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaModulePathPlugin.java

```diff
@@ -0,0 +1,194 @@
+    static final Set<String> EXCLUDES = Set.of(":test:framework", ":x-pack:plugin:eql:qa:common");
+
+    static void configureCompileModulePath(Project project) {
+        // first disable Gradle's builtin module path inference
+        project.getTasks()
+            .withType(JavaCompile.class)
+            .configureEach(compileTask -> compileTask.getModularity().getInferModulePath().set(false));
+
+        // test:framework has split pkgs with server, libs and more. do not use module path
+        var projName = project.toString();
+        if (EXCLUDES.stream().anyMatch(name -> projName.contains(name))) {
+            return;
+        }
+
+        var isModuleProject = hasModuleInfoDotJava(project);
+        var configurations = project.getConfigurations();
+        var compileClasspath = configurations.getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME);
+
+        var moduleCompileClasspath = configurations.create("moduleCompileClasspath", it -> {
+            it.extendsFrom(compileClasspath);
+            it.setCanBeConsumed(false); // We don't want this configuration used by dependent projects
+            it.attributes(
+                attrs -> attrs.attribute(
+                    LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE,
+                    project.getObjects().named(LibraryElements.class, LibraryElements.CLASSES)
+                )
+            );
+        }).getIncoming().artifactView(it -> {
+            it.componentFilter(cf -> {
+                var visited = new HashSet<ComponentIdentifier>();
+                return walkResolvedComponent(
+                    project,
+                    compileClasspath.getIncoming().getResolutionResult().getRoot(),
+                    isModuleProject,
+                    visited
+                ).anyMatch(cf::equals);
+            });
+        }).getFiles();
+
+        project.getTasks().named("compileJava", JavaCompile.class).configure(task -> {
+            var argumentProvider = new CompileModulePathArgumentProvider(isModuleProject, moduleCompileClasspath);
+            task.getOptions().getCompilerArgumentProviders().add(argumentProvider);
+            FileCollection classpath = task.getClasspath();
+            if (isIdea() == false && task.getClasspath() != null) {
+                FileCollection trimmedClasspath = classpath.minus(moduleCompileClasspath);
+                task.setClasspath(project.files(trimmedClasspath));
+            }
+            task.doLast(new Action<Task>() {
+                @Override
+                public void execute(Task task2) {
+                    // System.out.println("%s, Module path args: %s".formatted(project, argsToString(argumentProvider.asArguments())));
+                    // System.out.println("%s, Classpath: %s".formatted(project, pathToString(task.getClasspath().getAsPath())));
+                }
+            });
```

Comment:
Can be removed I guess

Reply:
I found this to be useful on numerous occasions, and having it in comments allowed me to quickly enable it.

Reply:
Or change this to use the task logger at info level?

Reply:
Great idea. Done.